### PR TITLE
LICM: use IRBuilder in hoist BO assoc

### PR DESCRIPTION
**Pull request by artagnon**
Use IRBuilder when creating the new invariant instruction, so that the constant-folder has an opportunity to constant-fold the new Instruction that we desire to create.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
The change to use IRBuilder looks good to me, but the rest seems like it will make generalization harder, not easier. You dropped some important checks that are only relevant for generalization, such as the need to check that both binops are associative (this is not implied by having the same opcode) or that the nuw transfer logic is only correct for adds.

**Comment by artagnon** [comment]
> The change to use IRBuilder looks good to me, but the rest seems like it will make generalization harder, not easier.

I was thinking we could first extend the same case to commutative ops: so `std::swap(LV, C1)` should suffice no?

> You dropped some important checks that are only relevant for generalization, such as the need to check that both binops are associative (this is not implied by having the same opcode) or that the nuw transfer logic is only correct for adds.

Fixed this thinko.

File: llvm/lib/Transforms/Scalar/LICM.cpp [hunk 1]
```diff
@@ -2816,40 +2816,43 @@ static bool hoistBOAssociation(Instruction &I, Loop &L,
...
   auto *NewBO = BinaryOperator::Create(
       Opcode, LV, Inv, BO->getName() + ".reass", BO->getIterator());
 
   // Copy NUW for ADDs if both instructions have it.
   if (Opcode == Instruction::Add && BO->hasNoUnsignedWrap() &&
       BO0->hasNoUnsignedWrap()) {
-    Inv->setHasNoUnsignedWrap(true);
+    // If the constant-folder didn't kick in, and a new Instruction was created.
+    if (auto *I = dyn_cast<Instruction>(Inv))
+      I->setHasNoUnsignedWrap(true);
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/LICM.cpp:2855]: It's better to bass the flag directly to CreateBinOp instead.

**Comment by nikic** [comment]
> > The change to use IRBuilder looks good to me, but the rest seems like it will make generalization harder, not easier.
> 
> I hope the PatternMatch change is fine.

It doesn't seem like an improvement, given that you still need to `cast<BinaryOperator>` for both afterwards.

File: llvm/lib/Transforms/Scalar/LICM.cpp (hunk 1 above)
Comment (artagnon) [reply llvm/lib/Transforms/Scalar/LICM.cpp:2855]: The `CreateBinOp` API doesn't seem to have a way to pass NUW/NSW flags: only other functions like `CreateAdd`/`CreateSub` seem to have it.

**Comment by artagnon** [comment]
> > > The change to use IRBuilder looks good to me, but the rest seems like it will make generalization harder, not easier.
> > 
> > 
> > I hope the PatternMatch change is fine.
> 
> It doesn't seem like an improvement, given that you still need to `cast<BinaryOperator>` for both afterwards.

Personally, I find a match to be more readable than multiple `getOperand(0)` and `getOperand(1)`, but if you feel this is not an improvement, I can drop this hunk.

File: llvm/lib/Transforms/Scalar/LICM.cpp [hunk 2]
```diff
@@ -2816,40 +2816,43 @@ static bool hoistBOAssociation(Instruction &I, Loop &L,
...
                                DominatorTree *DT) {
-  auto *BO = dyn_cast<BinaryOperator>(&I);
-  if (!BO || !BO->isAssociative())
-    return false;
+  using namespace PatternMatch;
 
-  // Only fold ADDs for now.
+  // Transform "(LV op C1) op C2" ==> "LV op (C1 op C2)"
+  Value *LV, *C1, *C2;
+  if (!match(&I, m_BinOp(m_BinOp(m_Value(LV), m_Value(C1)), m_Value(C2))) ||
```
Comment (artagnon) [inline llvm/lib/Transforms/Scalar/LICM.cpp:2823]: For the PatternMatch change: we can just turn this into `m_c_BinOp` to handle the other cases, no?

Comment (nikic) [reply llvm/lib/Transforms/Scalar/LICM.cpp:2823]: Still not convinced.

**Comment by nikic** [comment]
This PR still does a lot of unnecessary movement of conditions and style changes.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**